File: assets/Scripts/environment/sky/SkyClock.js
"use strict";

const SkyMath = require("./SkyMath.js");

class SkyClock {
    constructor() {
        this.enabled = true;

        // CDPR-like default: 20 minutes for a full cycle (day+night)
        this.dayLengthSec = 1200.0;

        this.t = 0.0;

        this._startApplied = false;

        // debug
        this._dbgPhase = -1; // 0..7
    }

    applyCfg(cfg) {
        if (!cfg) return;

        if (cfg.enabled === true) this.enabled = true;
        if (cfg.enabled === false) this.enabled = false;

        const dls = +cfg.dayLengthSec;
        if (Number.isFinite(dls) && dls > 1) this.dayLengthSec = dls;

        if (!this._startApplied && cfg.startTime01 != null) {
            const t01 = SkyMath.wrap(+cfg.startTime01, 0, 1);
            this.t = this.dayLengthSec * t01;
            this._startApplied = true;
        }

        if (ENGINE && ENGINE.log && ENGINE.log.debug) {
            ENGINE.log.debug(
                "[sky][clock] applyCfg enabled=" + this.enabled +
                " dayLengthSec=" + this.dayLengthSec +
                " startApplied=" + this._startApplied
            );
        }
    }

    step(dt) {
        if (!this.enabled) return;

        const d = Number.isFinite(+dt) ? +dt : 0.0;
        this.t += d;

        if (this.dayLengthSec > 0) {
            while (this.t > this.dayLengthSec) this.t -= this.dayLengthSec;
            while (this.t < 0) this.t += this.dayLengthSec;
        }

        // phase debug (8 steps) — cheap and useful
        const phase = Math.floor(this.time01 * 8);
        if (phase !== this._dbgPhase) {
            this._dbgPhase = phase;
            if (ENGINE && ENGINE.log && ENGINE.log.debug) {
                ENGINE.log.debug("[sky][clock] phase=" + phase + " time01=" + this.time01.toFixed(4));
            }
        }
    }

    setTime01(time01) {
        const t01 = SkyMath.wrap(+time01, 0, 1);
        this.t = this.dayLengthSec * t01;

        if (ENGINE && ENGINE.log && ENGINE.log.debug) {
            ENGINE.log.debug("[sky][clock] setTime01=" + t01.toFixed(4) + " tSec=" + this.t.toFixed(3));
        }
    }

    setTimeSec(timeSec) {
        const ts = +timeSec;
        if (!Number.isFinite(ts)) return;

        this.t = ts;

        if (this.dayLengthSec > 0) {
            while (this.t > this.dayLengthSec) this.t -= this.dayLengthSec;
            while (this.t < 0) this.t += this.dayLengthSec;
        }

        if (ENGINE && ENGINE.log && ENGINE.log.debug) {
            ENGINE.log.debug("[sky][clock] setTimeSec=" + ts.toFixed(3) + " time01=" + this.time01.toFixed(4));
        }
    }

    get time01() {
        return this.dayLengthSec > 0 ? (this.t / this.dayLengthSec) : 0;
    }
}

module.exports = SkyClock;

File: assets/Scripts/environment/sky/SkyClock.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local LuaNumber = luaRuntime.LuaNumber
local LuaNumberIsFinite = luaRuntime.LuaNumberIsFinite
local LuaNumberToFixed = luaRuntime.LuaNumberToFixed
local LuaDefineProperty = luaRuntime.LuaDefineProperty
SkyMath = require("./SkyMath.lua")
SkyClock = LuaClass()
SkyClock.name = "SkyClock"
function SkyClock.prototype.lua_constructor(self)
    self.enabled = true
    self.dayLengthSec = 1200
    self.t = 0
    self._startApplied = false
    self._dbgPhase = -1
end
function SkyClock.prototype.applyCfg(self, cfg)
    if not cfg then
        return
    end
    if cfg.enabled == true then
        self.enabled = true
    end
    if cfg.enabled == false then
        self.enabled = false
    end
    local dls = LuaNumber(cfg.dayLengthSec)
    if LuaNumberIsFinite(dls) and dls > 1 then
        self.dayLengthSec = dls
    end
    if not self._startApplied and cfg.startTime01 ~= nil then
        local t01 = SkyMath:wrap(
            LuaNumber(cfg.startTime01),
            0,
            1
        )
        self.t = self.dayLengthSec * t01
        self._startApplied = true
    end
    if ENGINE and ENGINE.log and ENGINE.log.debug then
        ENGINE.log:debug((((("[sky][clock] applyCfg enabled=" .. tostring(self.enabled)) .. " dayLengthSec=") .. tostring(self.dayLengthSec)) .. " startApplied=") .. tostring(self._startApplied))
    end
end
function SkyClock.prototype.step(self, dt)
    if not self.enabled then
        return
    end
    local lua_Number_isFinite_result_0
    if LuaNumberIsFinite(LuaNumber(dt)) then
        lua_Number_isFinite_result_0 = LuaNumber(dt)
    else
        lua_Number_isFinite_result_0 = 0
    end
    local d = lua_Number_isFinite_result_0
    self.t = self.t + d
    if self.dayLengthSec > 0 then
        while self.t > self.dayLengthSec do
            self.t = self.t - self.dayLengthSec
        end
        while self.t < 0 do
            self.t = self.t + self.dayLengthSec
        end
    end
    local phase = math.floor(self.time01 * 8)
    if phase ~= self._dbgPhase then
        self._dbgPhase = phase
        if ENGINE and ENGINE.log and ENGINE.log.debug then
            ENGINE.log:debug((("[sky][clock] phase=" .. tostring(phase)) .. " time01=") .. LuaNumberToFixed(self.time01, 4))
        end
    end
end
function SkyClock.prototype.setTime01(self, time01)
    local t01 = SkyMath:wrap(
        LuaNumber(time01),
        0,
        1
    )
    self.t = self.dayLengthSec * t01
    if ENGINE and ENGINE.log and ENGINE.log.debug then
        ENGINE.log:debug((("[sky][clock] setTime01=" .. tostring(LuaNumberToFixed(t01, 4))) .. " tSec=") .. LuaNumberToFixed(self.t, 3))
    end
end
function SkyClock.prototype.setTimeSec(self, timeSec)
    local timeValue = LuaNumber(timeSec)
    if not LuaNumberIsFinite(timeValue) then
        return
    end
    self.t = timeValue
    if self.dayLengthSec > 0 then
        while self.t > self.dayLengthSec do
            self.t = self.t - self.dayLengthSec
        end
        while self.t < 0 do
            self.t = self.t + self.dayLengthSec
        end
    end
    if ENGINE and ENGINE.log and ENGINE.log.debug then
        ENGINE.log:debug((("[sky][clock] setTimeSec=" .. LuaNumberToFixed(timeValue, 3)) .. " time01=") .. LuaNumberToFixed(self.time01, 4))
    end
end
LuaDefineProperty(
    SkyClock.prototype,
    "time01",
    {get = function(self)
        local lua_temp_1
        if self.dayLengthSec > 0 then
            lua_temp_1 = self.t / self.dayLengthSec
        else
            lua_temp_1 = 0
        end
        return lua_temp_1
    end},
    true
)
M = SkyClock

return M
